Scoped storage bugfixes

lrq3000 opened the pull request:
# Description

Fix residual issues after PR #32 merge:
* Read and write of filesync file now works as intended (before it always looked for a "filePath" file, instead of the path pointed by the `$filePath` variable). In the previous PR, without the bugfix, the user's specified filePath was respected the first time, on file creation, but afterwards the file was not updated nor read.
* Calls `grantFilePermissionCallBack()` after folder is picked and permission is granted. ~Web app needs an update too to call `isGrantedFilePermission()` only after the requestId `rId` is deleted from `requestMap` in `grantFilePermissionCallBack()`, as discussed in more details here: https://github.com/johannesjo/super-productivity-android/pull/32#issuecomment-1382965427 .~
* Implements `allowedFolderPath()` to display in the UI the currently granted path to avoid user confusion (@zoli's sugestion).
* A few bugfixes here and there for edge cases, especially in `readFile()`.
* Updated README.md instructions to access local webserver from Android emulator.

A debug APK is available here: ~https://github.com/lrq3000/super-productivity-android/releases/tag/super-productivity-android_scoped-storage_bugfix1~ ~https://github.com/lrq3000/super-productivity-android/releases/tag/super-productivity-android_scoped-storage_bugfix2~ https://github.com/lrq3000/super-productivity-android/releases/tag/super-productivity-android_scoped-storage_bugfix3

## Issues Resolved

Fixes #13 (again 😅), fixes https://github.com/johannesjo/super-productivity/issues/2001 , fixes https://github.com/johannesjo/super-productivity/issues/690 .

## Check List

- [x] ~Web app update to check `isGrantedFilePermission()` after `grantFilePermissionCallBack()`. Until then, the app must be killed and relaunched after folder selection, or the folder needs to be selected twice, to force the app to query again the permissions and hence be able to Enable Syncing (otherwise we get a "Needs permission" warning message).~ Fixed `requestCode` into `requestId` in `grantFilePermissionCallBack()` and it now works fine!
- [x] Implement `allowedFolderPath()` to display in the UI the currently granted path to avoid user confusion (@zoli's sugestion).
- [x] Tested bidirectional syncing myself (via Syncthing-Fork on Android + Synctrayzor on Windows 10).
- [x] Someone else needs to test too.

lrq3000 committed:
callback grantFilePermissionCallback() to avoid concurrency issues (fixes need to select folder twice)

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
fix filepath issues for filesync feature
(before it always pointed to a filePath file, instead of variable $filePath)

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
remove unnecessary callback (thanks to @anggrayudi !)

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 commented:
This time around I know a bit more how to use the software, I think file syncing works as expected now.

But still I'm pretty new to Super Productivity, it would be great if someone else could try too to double check everything is working as expected.

zoli commented:
Thanks @lrq3000. I tested this and just wanted to suggest STH.
Other than the mentioned issue with the callback, I think its better to mention that the path is relative in the UI and it will be better if we could show the it is relative to what dir, because its probable for the user to forget which directory he/she had given permission.
I myself just used my old path and it resulted in creating the whole path under the relative directory that I gave permission to read and write.

lrq3000 commented:
@zoli Thank you for your feedback! Yes showing the current path would be nice. If someone like @johannesjo can help with the typescript part, I can modify the callback functions such as isGrantedFilePermission to return the folder path, and on the webapp this can then be displayed in a non-editable text label.

I also tried an alternative approach, to allow the user to select a file instead of a folder, so that we could directly store the full filepath instead of the folder and filepath separately, but the way file picking works in Android currently does not allow to do this in a straightforward manner, we can only pick already existing files. So I dropped this idea.

johannesjo commented:
Thank you two!!

I still get an error when I click on sync with either the existing or a completely new sync file path:
```
2023-01-16 14:56:50.258 3907-4034/com.superproductivity.superproductivity D/SuperProductivity: readFile error: java.lang.StringIndexOutOfBoundsException: String index out of range: -1
        at java.lang.AbstractStringBuilder.deleteCharAt(AbstractStringBuilder.java:824)
        at java.lang.StringBuilder.deleteCharAt(StringBuilder.java:254)
        at com.superproductivity.superproductivity.CommonJavaScriptInterface.readFile(CommonJavaScriptInterface.kt:384)
        at android.os.MessageQueue.nativePollOnce(Native Method)
        at android.os.MessageQueue.next(MessageQueue.java:336)
        at android.os.Looper.loop(Looper.java:174)
        at android.os.HandlerThread.run(HandlerThread.java:67)
```
It's also worth noting that the error get's kind of swallowed (from the web frontend's perspective). Maybe it's better to ditch the `try .. catch` blocks?

@lrq3000 Could you also outline more specifically what changes we need to the file permission interface?

**EDIT** [Just saw your comment ](https://github.com/johannesjo/super-productivity-android/pull/32#issuecomment-1382965427)

johannesjo commented:
@lrq3000 

> Hence, isGrantedFilePermission() is called right away after grantFilePermission() returns. This worked before because grantFilePermission() used to directly request for permission in a blocking manner in older Android versions, but now grantFilePermission() calls an asynchronous folder/file picker which itself calls back storageHelper.onFolderSelected() (and other internal functions to storageHelper) which saves the permission.

This is a bit misleading. The functions are not called there (the round braces are missing). This is a check to see if those two functions are defined on the android interface to determine if file sync is available as a feature. The actual call to grantFilePermission() is made a few lines below and is only triggered when the file sync provider is selected:
```
        change: (field) => {
          if (
            IS_ANDROID_WEB_VIEW &&
            field.model.syncProvider === SyncProvider.LocalFile
          ) {
            androidInterface.grantFilePermissionWrapped().then(() => {
              field.formControl?.updateValueAndValidity();
            });
          }
        },
```
`isGrantedFilePermission` is in fact never called. 

What would be the correct order of steps?

lrq3000 committed:
Remove unnecessary last line return deletion in readFile()

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
more idiomatic code

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
simplify readFile(), no need to replace line return character

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 commented:
@johannesjo About the exception you encountered, this is not due to code I added but code that was already there. I'm not sure why but the readFile() function reads the filesync file and then replaces each line return with the system's line return character, and then deletes the last, unnecessary line return - this last bit is what caused the exception for you, because at first the file does not exist, so trying to read it may lead to this issue. I have now removed this logic altogether, we simply read the file's content as a String, and from my own tests it seems to work well.

About `isGrantedFilePermission()`, sorry I'm not good with typescript :-/ What is needed is that the webapp calls `isGrantedFilePermission()` only after `grantFilePermissionCallBack()` gets called.

Also if we want to implement @zoli's suggestion, I can modify `isGrantedFilePermission()` to return a String, the path to the folder or null if no permission, instead of just a Boolean. You can then use this String in the webapp to display in a field or text label.

lrq3000 commented:
@johannesjo also if you still experience an exception, can  you please let me know what android version you are running? Because i don't get these exceptions, i guess you are using another version of android with a slightly different file management api.

I also didn't notice you said that isGrantedFilePermission is never called, but it is necessarily. I can see it being called in the Android app, and it's not called from within the app, it necessarily is called from the typescript webapp somehow. If you really have no idea what is going on, I'll try to trace the whole process later today.

johannesjo commented:
I am still encountering errors:
1. When I tried to use the already existing file this happened:
```
2023-01-17 14:33:56.448 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 1 (1) 7 4
2023-01-17 14:33:56.450 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 7 8
2023-01-17 14:33:56.490 7459-7562/com.superproductivity.superproductivity D/SuperProductivity: writeFile: trying to save to filePath: Hh2hh.json
2023-01-17 14:33:56.490 7459-7562/com.superproductivity.superproductivity D/SuperProductivity: writeFile: trying to save to fullFilePath: /storage/emulated/0/Download/Hh2hh.json
2023-01-17 14:33:56.520 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 1 (1) 7 4
2023-01-17 14:33:56.522 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 7 8
2023-01-17 14:33:56.542 7459-7562/com.superproductivity.superproductivity D/SuperProductivity: writeFile: erase file content by recreating it
2023-01-17 14:33:56.551 7459-7562/com.superproductivity.superproductivity D/SuperProductivity: writeFile: try to openOutputStream
2023-01-17 14:33:56.551 7459-7562/com.superproductivity.superproductivity W/System.err: java.lang.NullPointerException
2023-01-17 14:33:56.551 7459-7562/com.superproductivity.superproductivity W/System.err:     at com.superproductivity.superproductivity.CommonJavaScriptInterface.writeFile(CommonJavaScriptInterface.kt:422)
2023-01-17 14:33:56.551 7459-7562/com.superproductivity.superproductivity W/System.err:     at android.os.MessageQueue.nativePollOnce(Native Method)
2023-01-17 14:33:56.551 7459-7562/com.superproductivity.superproductivity W/System.err:     at android.os.MessageQueue.next(MessageQueue.java:336)
2023-01-17 14:33:56.551 7459-7562/com.superproductivity.superproductivity W/System.err:     at android.os.Looper.loop(Looper.java:174)
2023-01-17 14:33:56.551 7459-7562/com.superproductivity.superproductivity W/System.err:     at android.os.HandlerThread.run(HandlerThread.java:67)
2023-01-17 14:33:56.555 7459-7562/com.superproductivity.superproductivity W/TW: JavascriptInterface: updateNotificationWidget
2023-01-17 14:33:56.559 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 1 (1) 7 4
2023-01-17 14:33:56.560 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 7 8
2023-01-17 14:33:56.561 7459-7459/com.superproductivity.superproductivity I/chromium: [INFO:CONSOLE(1)] "__error during sync__", source: https://app.super-productivity.com/main.05ec82a44189ec80.js (1)
2023-01-17 14:33:56.561 7459-7459/com.superproductivity.superproductivity I/chromium: [INFO:CONSOLE(1)] "Error: Error: Java exception was raised during method invocation", source: https://app.super-productivity.com/main.05ec82a44189ec80.js (1)
```

2. When I tried to use a new file name this happened:
```
2023-01-17 14:35:07.098 7459-7562/com.superproductivity.superproductivity D/SuperProductivity: writeFile: file does not exist, try to create it
2023-01-17 14:35:07.105 7459-7562/com.superproductivity.superproductivity D/SuperProductivity: writeFile: do we have access to parentFolder? androidx.documentfile.provider.TreeDocumentFile@97d9160
2023-01-17 14:35:07.110 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 1 (1) 7 4
2023-01-17 14:35:07.111 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 7 8
2023-01-17 14:35:07.126 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 1 (1) 7 4
2023-01-17 14:35:07.127 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 7 8
2023-01-17 14:35:07.143 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 1 (1) 7 4
2023-01-17 14:35:07.144 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 7 8
2023-01-17 14:35:07.161 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 1 (1) 7 4
2023-01-17 14:35:07.162 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 7 8
2023-01-17 14:35:07.177 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 1 (1) 7 4
2023-01-17 14:35:07.177 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 7 8
2023-01-17 14:35:07.193 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 1 (1) 7 4
2023-01-17 14:35:07.194 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 7 8
2023-01-17 14:35:07.211 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 1 (1) 7 4
2023-01-17 14:35:07.212 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 7 8
2023-01-17 14:35:07.224 7459-7562/com.superproductivity.superproductivity D/SuperProductivity: writeFile: erase file content by recreating it
2023-01-17 14:35:07.226 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 1 (1) 7 4
2023-01-17 14:35:07.227 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 7 8
2023-01-17 14:35:07.241 7459-7562/com.superproductivity.superproductivity D/SuperProductivity: writeFile: try to openOutputStream
2023-01-17 14:35:07.242 7459-7562/com.superproductivity.superproductivity W/System.err: java.lang.NullPointerException
2023-01-17 14:35:07.244 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 1 (1) 7 4
2023-01-17 14:35:07.244 7459-7562/com.superproductivity.superproductivity W/System.err:     at com.superproductivity.superproductivity.CommonJavaScriptInterface.writeFile(CommonJavaScriptInterface.kt:422)
2023-01-17 14:35:07.244 7459-7562/com.superproductivity.superproductivity W/System.err:     at android.os.MessageQueue.nativePollOnce(Native Method)
2023-01-17 14:35:07.244 7459-7562/com.superproductivity.superproductivity W/System.err:     at android.os.MessageQueue.next(MessageQueue.java:336)
2023-01-17 14:35:07.244 7459-7562/com.superproductivity.superproductivity W/System.err:     at android.os.Looper.loop(Looper.java:174)
2023-01-17 14:35:07.244 7459-7562/com.superproductivity.superproductivity W/System.err:     at android.os.HandlerThread.run(HandlerThread.java:67)
2023-01-17 14:35:07.244 7459-7491/com.superproductivity.superproductivity D/eglCodecCommon: setVertexArrayObject: set vao to 0 (0) 7 8
2023-01-17 14:35:07.245 7459-7459/com.superproductivity.superproductivity I/chromium: [INFO:CONSOLE(1)] "__error during sync__", source: https://app.super-productivity.com/main.05ec82a44189ec80.js (1)
2023-01-17 14:35:07.246 7459-7459/com.superproductivity.superproductivity I/chromium: [INFO:CONSOLE(1)] "Error: Error: Java exception was raised during method invocation", source: https://app.super-productivity.com/main.05ec82a44189ec80.js (1)
```

 



Currently `isGrantedFilePermission` is never called, so nothing needs to be changed :)

I use the `is` to indicate boolean functions so maybe we can rather provide a new interface, something like `allowedFolderPath`. To understand this better: Is it always just a single folder that the app can have access to? And we don't need to pass it from the web frontend as we know the value in the android world, right?

lrq3000 commented:
@johannesjo thank you for the logs, it seems you don't have permission to the file for some reason, but you do have permission on a folder, but it's unclear which one. Could you please tell me which version of Android you are running? Is it in an emulator or real device? I tested on both an emulador and a real device and it works perfectly well so at this point i will need to reproduce your environment to be able to debug.

About allowedFolderPath, yes it is stored in the android app. We can store either a path to an existing folder or an existing file, but Android does not provide a native dialog to pick a non existing file, so that's why i can only show a dialog to pick a folder. So yes i can return the folder path to the webapp, but it is stored anyway. I could even store the path to the file too, but it needs to be entered in a textbox as it is now, i can't provide a dialog to pick a non exsting file.

About isGrantedFilePermission, i will test again, but I'm pretty sure I already tested and this is the function that the webapp calls somewhere (not where i pointed, there are ither places) to check if it has permission to enable filesync.

johannesjo commented:
I was finally able to get it to work. I had to switch away from the file sync provider and switch to it again and click through the allow dialog multiple times, but I now it works (maybe it's possible to polish this a bit later).

> About isGrantedFilePermission, i will test again, but I'm pretty sure I already tested and this is the function that the webapp calls somewhere (not where i pointed, there are ither places) to check if it has permission to enable filesync.

Sorry! My bad! I got confused. It's actually called multiple times. I will make the adjustments you suggested (I hope I get to it tomorrow).

lrq3000 commented:
Oh ok great, thank you! Yes the issue with selecting the folder multiple times before it works will be fixed once you implement the isGrantedFilePermission fix I mentioned above, then it should work on first try.

One major reason why I can't modify the web app myself, beyond my lack of typescript skills, is that I cannot access the local angular dev server that is run with `ng serve` from the Android Studio emulator. Do you maybe have some tips? It's only an issue with the Angular server, because if I run another server, like Xampp, then it can be accessed without any issue from the Android emulator. It seems the angular server is not serving other clients than localhost for some reason, and `--disable-host-check` does not help.

lrq3000 committed:
update instructions and config to access host's local webserver from inside Android's emulator

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 committed:
restore default debug url to local webserver

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 commented:
Ok after **hours** of cumbersome debugging, I could find a way to make the locally served web app accessible from inside the Android emulator. I have updated the instructions in the README.

Still, there is an issue. The web app launches, but it stays in infinite loading, but only from within the Android emulator, not from the host's web browser.

lrq3000 committed:
implement allowedFolderPath()

Signed-off-by: Stephen L. <lrq3000@gmail.com>

lrq3000 commented:
@johannesjo I have implemented `allowedFolderPath()`. It returns 3 possible String values:
* either a path to a folder on android 10+ if we have permission
* either `"<any>"` on android 9- if we have permission over the whole filesystem
* either `""` (empty string) on any android version if we don't have permission

Please let me know if this is OK or if you would like some change.

johannesjo commented on app/src/main/java/com/superproductivity/superproductivity/CommonJavaScriptInterface.kt:
```diff
@@ -439,13 +499,17 @@ abstract class CommonJavaScriptInterface(
                     sp.edit().putString("filesyncFolder", fpath).apply()
+                    // Once permissions are granted, callback web application to continue execution
+                    callJavaScriptFunction(
+                        FN_PREFIX + "grantFilePermissionCallBack('" + requestCode + "')"
```
```suggestion
                        FN_PREFIX + "grantFilePermissionCallBack('" + requestId + "')"
```

lrq3000 replied:
Thank you, missed this typo caused by a copy/paste, it's now merged!

johannesjo commented:
Thank you very much for your efforts @lrq3000 and sorry about having to spend so much time on the web app setup. I should have written the part to the readme myself... 

I added showing the allowedFolderPath. I also found a little mistake in the android code (see above) that was responsible for the issue we discussed above as the web-frontend was never able to resolve the promise of `grantFilePermission`.

I think everything should work now!! If you have no further remarks, I gladly will merge this!

lrq3000 committed:
fix requestId in callback

Co-authored-by: Johannes Millan <johannesjo@users.noreply.github.com>

lrq3000 commented:
Awesome, thank you @johannesjo for your responsiveness in following up on this issue :-) And good catch about `requestId`, I thought I double checked but this bug slipped through! I tested again and indeed folder picking works the first time around now :-)

I have no further remarks. I will try to stay available if other issues with this feature pop up, but from my own tests it seems to work fine (I am now already using Super Productivity as part of my daily workflow using my debug builds :D).

johannesjo commented:
Great! I am merging this then. I plan to make a new android release at the start of February which will remove GoogleDrive support (since it doesn't work any more). This way users have a better alternative at hand :)

Thank you very much again @lrq3000 and also @zoli for your input!!!! <3

lrq3000 commented:
Awesome, thank you so much @johannesjo for your active support and reviewing of this PR! I expect the submission process to the Google Play Store should go without a hitch now that Super Productivity does not use the `MANAGE_EXTERNAL_STORAGE` anymore :-)
